async_grpc/client.h

```diff
@@ -84,119 +77,221 @@ class Client {
+  void Reset() {
+    client_context_ = common::make_unique<::grpc::ClientContext>();
+  }
+
+  bool WriteImpl(const RequestType &request, ::grpc::Status *status) {
+    auto status_normal_rpc = MakeBlockingUnaryCall(request, &response_);
+    if (status != nullptr) {
+      *status = status_normal_rpc;
     }
+    return status_normal_rpc.ok();
+  }
+  ::grpc::Status MakeBlockingUnaryCall(const RequestType &request,
+                                       ResponseType *response) {
+    return ::grpc::internal::BlockingUnaryCall(
+        channel_.get(), rpc_method_, client_context_.get(), request, response);
   }
 
-  ::grpc::Status StreamFinish() {
-    switch (rpc_method_.method_type()) {
-      case ::grpc::internal::RpcMethod::CLIENT_STREAMING:
-        InstantiateClientWriterIfNeeded();
-        return client_writer_->Finish();
-      case ::grpc::internal::RpcMethod::BIDI_STREAMING:
-        InstantiateClientReaderWriterIfNeeded();
-        return client_reader_writer_->Finish();
-      case ::grpc::internal::RpcMethod::SERVER_STREAMING:
-        CHECK(client_reader_);
-        return client_reader_->Finish();
-      default:
-        LOG(FATAL) << "This method is for streaming RPC only.";
+  std::shared_ptr<::grpc::Channel> channel_;
+  std::unique_ptr<::grpc::ClientContext> client_context_;
+  const std::string rpc_method_name_;
+  const ::grpc::internal::RpcMethod rpc_method_;
+
+  ResponseType response_;
+  RetryStrategy retry_strategy_;
+};
+
+template <typename RpcServiceMethodConcept>
+class Client<RpcServiceMethodConcept,
+             ::grpc::internal::RpcMethod::CLIENT_STREAMING> {
+  using RpcServiceMethod = RpcServiceMethodTraits<RpcServiceMethodConcept>;
+  using RequestType = typename RpcServiceMethod::RequestType;
+  using ResponseType = typename RpcServiceMethod::ResponseType;
+
+ public:
+  Client(std::shared_ptr<::grpc::Channel> channel)
+      : channel_(channel),
+        client_context_(common::make_unique<::grpc::ClientContext>()),
+        rpc_method_name_(RpcServiceMethod::MethodName()),
+        rpc_method_(rpc_method_name_.c_str(), RpcServiceMethod::StreamType,
+                    channel_) {}
+
+  bool Write(const RequestType &request, ::grpc::Status *status = nullptr) {
+    ::grpc::Status internal_status;
+    WriteImpl(request, &internal_status);
+    if (status != nullptr) {
+      *status = internal_status;
     }
+    return internal_status.ok();
+  }
+
+  bool StreamWritesDone() {
+    InstantiateClientWriterIfNeeded();
+    return client_writer_->WritesDone();
   }
 
-  const ResponseType &response() {
-    CHECK(rpc_method_.method_type() ==
-              ::grpc::internal::RpcMethod::NORMAL_RPC ||
-          rpc_method_.method_type() ==
-              ::grpc::internal::RpcMethod::CLIENT_STREAMING);
-    return response_;
+  ::grpc::Status StreamFinish() {
+    InstantiateClientWriterIfNeeded();
+    return client_writer_->Finish();
   }
 
+  const ResponseType &response() { return response_; }
```

Comment:
super-nit: ` const ResponseType& response() `to be consistent with the rest of code base. Clang-format can change it back though, if it finds another ref styled like that first.

Reply:
Done